src/LanguageServer/ProtocolUnitTests/Definitions/GoToTypeDefinitionTests.cs

```diff
@@ -81,6 +156,96 @@ class B
+                        public B b{|caret:|};
+                    }
+                }
+                """;
+            var mapped =
+                """
+                namespace M
+                {
+                    class B
+                    {
+                    }
+                }
+                """;
+
+            await using var testLspServer = await CreateTestLspServerAsync(source, mutatingLspWorkspace);
+
+            AddMappedDocument(testLspServer.TestWorkspace, mapped);
+
+            var results = await RunGotoTypeDefinitionAsync(testLspServer, testLspServer.GetLocations("caret").Single());
+            var result = Assert.Single(results);
+            AssertLocationsEqual([TestSpanMapper.MappedFileLocation], results);
+        }
+
+        [Theory, CombinatorialData]
+        public async Task TestGotoTypeDefinitionAsync_SourceGeneratedDocument(bool mutatingLspWorkspace)
+        {
+            var source =
+                """
+                namespace M
+                {
+                    class A
+                    {
+                        public B b{|caret:|};
+                    }
+                }
+                """;
+            var generated =
+                """
+                namespace M
+                {
+                    class B
+                    {
+                    }
+                }
+                """;
+
+            await using var testLspServer = await CreateTestLspServerAsync(source, mutatingLspWorkspace);
+            await AddGeneratorAsync(new SingleFileTestGenerator(generated), testLspServer.TestWorkspace);
+
+            var results = await RunGotoTypeDefinitionAsync(testLspServer, testLspServer.GetLocations("caret").Single());
+            var result = Assert.Single(results);
+            Assert.Equal(SourceGeneratedDocumentUri.Scheme, result.Uri.Scheme);
+        }
+
+        [Theory, CombinatorialData]
+        public async Task TestGotoTypeDefinitionAsync_CrossLanguage(bool mutatingLspWorkspace)
+        {
+            var markup =
+@"<Workspace>
+    <Project Language=""C#"" Name=""Definition"" CommonReferences=""true"" FilePath=""C:\CSProj1.csproj"">
+        <Document FilePath=""C:\A.cs"">
+            public class {|definition:A|}
+            {
+            }
+        </Document>
+    </Project>
+    <Project Language=""Visual Basic"" CommonReferences=""true"" FilePath=""C:\CSProj2.csproj"">
+        <ProjectReference>Definition</ProjectReference>
+        <Document FilePath=""C:\C.cs"">
+            Class C
+                Dim {|caret:a|} As A
+            End Class
+        </Document>
+    </Project>
+</Workspace>";
+            await using var testLspServer = await CreateXmlTestLspServerAsync(markup, mutatingLspWorkspace);
+
+            var results = await RunGotoTypeDefinitionAsync(testLspServer, testLspServer.GetLocations("caret").Single());
+            AssertLocationsEqual(testLspServer.GetLocations("definition"), results);
+        }
```

Comment:
nice tests! can we potentially add one for metadata?  Some examples here in case you need it - https://github.com/dotnet/roslyn/blob/main/src/LanguageServer/ProtocolUnitTests/Metadata/LspMetadataAsSourceWorkspaceTests.cs